torch/sparse/semi_structured.py

```diff
@@ -0,0 +1,318 @@
+            elif isinstance(input_B, cls) and input_B.transposed:
+                res = input_B.t().cslt.addmm(input_A.T, bias, True, cls._fuse_transpose)  # type: ignore[attr-defined]
+                return res if cls._fuse_transpose else res.T
+
+            raise NotImplementedError(
+                (
+                    f"func: {func} is currently not supported for ",
+                    f"opA: {input_A.transposed} A: {input_A}",
+                    f"opB: {input_B.transposed} B {input_B}",
+                )
+            )
+
+        if func is torch.ops.aten.mm.default:
+            input_A, input_B = args
+
+            if (
+                isinstance(input_A, cls)
+                and not input_A.transposed
+            ):
+                return input_A.cslt.mm(input_B, not input_B.is_contiguous(), False)  # type: ignore[attr-defined]
+
+            elif isinstance(input_B, cls) and input_B.transposed:
+                res = input_B.t().cslt.mm(input_A.T, True, cls._fuse_transpose)  # type: ignore[attr-defined]
+                return res if cls._fuse_transpose else res.T
+
+            raise NotImplementedError(
+                (
+                    f"func: {func} is currently not supported for ",
+                    f"opA: {input_A.transposed} A: {input_A}",
+                    f"opB: {input_B.transposed} B {input_B}",
+                )
+            )
+
+        # When torch is run with inference mode, it looks like pytorch does some merging in order to make linear faster.
+        # The end result is that it will use this aten.linear.default op instead of decomposing into a .t() and addmm()
+        # We handle this case in order to support with torch.inference_mode()
+        if func is torch.ops.aten.linear.default:
+            input, weight, bias = args
+            if isinstance(weight, cls):
+                res = weight.t().cslt.addmm(input.T, bias, True, cls._fuse_transpose)  # type: ignore[attr-defined]
+                return res if cls._fuse_transpose else res.T
+
+        # handle values
+        if func is torch.ops.aten.values.default:
+            m, k = args[0].shape
+            num_kept_elements = m * k // 2
+            return args[0].compressed_tensor[:num_kept_elements].view(m, k // 2)
+
+        # handle indices
+        if func is torch.ops.aten.indices.default:
+            m, k = args[0].shape
+            num_kept_elements = m * k // 2
+            return args[0].compressed_tensor[num_kept_elements:].view(m, -1).view(torch.int16)
+
+        raise NotImplementedError(f"{func} on {args} is not implemented!")
+
+
+def to_sparse_semi_structured(
+    original_tensor: torch.Tensor,
+    transposed=False,
+) -> SparseSemiStructuredTensor:
+    """
+    This function converts a dense tensor into a sparse semi-structured tensor.
+    It will return a SparseSemiStructuredTensor, a subclass of torch.Tensor. This subclass is
+    responsible for overriding __torch_dispatch__ for accelerated sparse matmul.
```

Comment:
nit: i'm not sure implementation details are necessary here beyond you get a Tensor subclass that will be able to leverage this sparsity.